### PR TITLE
[percep] Added y axis pass through filter closes #501

### LINKED ISSUE #501: Y axis pass through filter
Testing setup: Choose a folder to read in images from file I made for the project and log the output using >. Do this two times and diff the documents to make sure we can generate consistent output. Every time you make changes you can then compare them to your previous changes to see how things changed. You can also just look at video output if you'd rather not do this.

3. Edit the passthrough filter function to add a Y axis pass through filter, so we can discard points that are above a certain height. We don't want to be detecting tree branches as obstacles. Record the performance before and after this change. I'd say a good cutoff would probably be 2 meters off of the ground. Make sure that you are editing the current passthrough function and passing in the parameters that will change for the filtering, as opposed to creating an entirely new function.

### PR DESCRIPTION
**Pull request by JayeshM29**
Tested using the pcl images from perception-files folder. Tuned the y value upper limit until a small part of the hill top was cut off.

### CODE REVIEW

**Review by mmartin4972 — DISMISSED** [review]

File: jetson/percep/pcl.cpp [hunk 1]
```diff
@@ -6,16 +6,16 @@
 //Filters out all points with z values that aren't within a threshold
 //Z values are depth values in mm
 //Source: https://rb.gy/kkyi80
```
Comment (amverni) [inline jetson/percep/pcl.cpp:8]: Looks like this comment is now out of date

File: jetson/percep/pcl.cpp [hunk 2]
```diff
@@ -6,16 +6,16 @@
...
-void PCL::PassThroughFilter() {
+void PCL::PassThroughFilter(std::string axis, double upperLimit) {
```
Comment (amverni) [inline jetson/percep/pcl.cpp:9]: Can the arguments be made const?

File: jetson/percep/pcl.cpp [hunk 3]
```diff
@@ -480,4 +481,4 @@ void PCL::update() {
     std::cerr << "Height: "<< pt_cloud_ptr->height<<"\n";
 }
 
-#endif
+#endif
```
Comment (amverni) [inline jetson/percep/pcl.cpp:484]: add back the newline please (https://stackoverflow.com/questions/72271/no-newline-at-end-of-file-compiler-warning)

**Review by amverni — CHANGES_REQUESTED** [review]

File: jetson/percep/pcl.cpp [hunk 4]
```diff
@@ -458,14 +458,15 @@ shared_ptr<pcl::visualization::PCLVisualizer> PCL::createRGBVisualizer() {
 //This function is called in main.cpp
 obstacle_return PCL::pcl_obstacle_detection(shared_ptr<pcl::visualization::PCLVisualizer> viewer) {
     obstacle_return result;
-    PassThroughFilter();
+    PassThroughFilter("z", 7000.0);
+    PassThroughFilter("y", 3000.0);
```
Comment (raytitan) [inline jetson/percep/pcl.cpp:462]: How were these determined?

Comment (mmartin4972) [reply jetson/percep/pcl.cpp:462]: We've established that we can trust the ZED depth for up to 7 m which is where the 7k comes from. 3000 was just a placeholder. When we do more outdoor testing will tune that parameter

Comment (amverni) [reply jetson/percep/pcl.cpp:462]: 1. You should comment this.
2. This should be in a more convenient location such as a config file.

Comment (mmartin4972) [reply jetson/percep/pcl.cpp:462]: We have a config file in the pipeline

Comment (amverni) [reply jetson/percep/pcl.cpp:462]: I don't know what you mean by "pipeline" but maybe this means my comment should be "this should be in a more convenient locations such as your config file".

**Review by theonlynick0430 — APPROVED** [review]

File: jetson/percep/pcl.cpp (hunk 4 above)
Comment (mmartin4972) [reply jetson/percep/pcl.cpp:462]: I mean we're working on it. It's in the works

Comment (amverni) [reply jetson/percep/pcl.cpp:462]: Ahh got it. Then let's do this:

1. Go with the comment for now
2. Include a link to that issue in this PR (I.e. followup on this thread)
3. In that issue, add this item to the description so it is not forgotten.

Comment (mmartin4972) [reply jetson/percep/pcl.cpp:462]: Link to issue: https://github.com/umrover/mrover-workspace/issues/474

**Review by mmartin4972 — DISMISSED** [review]

File: jetson/percep/pcl.cpp [hunk 5]
```diff
@@ -3,19 +3,19 @@
 
 #if OBSTACLE_DETECTION
 /* --- Pass Through Filter --- */
-//Filters out all points with z values that aren't within a threshold
-//Z values are depth values in mm
+//Filters out all points on a given axis that aren't within a threshold
+//Values are depth values in mm
 //Source: https://rb.gy/kkyi80
-void PCL::PassThroughFilter() {
+void PCL::PassThroughFilter(const std::string axis, const double upperLimit) {
```
Comment (amverni) [inline jetson/percep/pcl.cpp:9]: It's not clear what values I can and should pass in for `axis`. You should at very least add this to the function comment. I think it might also be worth doing some error checking.

File: jetson/percep/pcl.cpp (hunk 4 above)
Comment (amverni) [reply jetson/percep/pcl.cpp:462]: Don't forget a comment!

File: jetson/percep/pcl.cpp [hunk 6]
```diff
@@ -480,4 +481,5 @@ void PCL::update() {
     std::cerr << "Height: "<< pt_cloud_ptr->height<<"\n";
 }
 
+
```
Comment (amverni) [inline jetson/percep/pcl.cpp:484]: NIT: Why was this added?

File: jetson/percep/pcl.hpp [hunk 7]
```diff
@@ -52,7 +52,7 @@ class PCL {
     private:
 
         //Filters points with values beyond certain threshold
-        void PassThroughFilter();
+        void PassThroughFilter(const std::string axis, const double upperLimit);
```
Comment (amverni) [inline jetson/percep/pcl.hpp:55]: `passThroughFilter` (lower case "p")? Might be worth a general cleanup of the entire CV project (in a separate issue, not this one).

File: jetson/percep/pcl.cpp [hunk 8]
```diff
@@ -3,19 +3,19 @@
 
 #if OBSTACLE_DETECTION
 /* --- Pass Through Filter --- */
-//Filters out all points with z values that aren't within a threshold
-//Z values are depth values in mm
+//Filters out all points on a given axis that aren't within a threshold
+//Values are depth values in mm
 //Source: https://rb.gy/kkyi80
```
Comment (amverni) [inline jetson/percep/pcl.cpp:8]: What if we made this one multiline comment instead of a mix of multiline and multiple one liners. I.e.:
```
/* Pass Through Filter
 * Filters out ...
 * ...
 */

**Review by amverni — CHANGES_REQUESTED** [review]
This is really coming along! Just a few more comments.

File: jetson/percep/pcl.cpp (hunk 8 above)
Comment (mmartin4972) [reply jetson/percep/pcl.cpp:8]: We've formatted all of our code with the current syntax

File: jetson/percep/pcl.hpp (hunk 7 above)
Comment (mmartin4972) [reply jetson/percep/pcl.hpp:55]: We'll address that in another issue

File: jetson/percep/pcl.cpp (hunk 4 above)
Comment (mmartin4972) [reply jetson/percep/pcl.cpp:462]: I thot I did?

**Review by mmartin4972 — APPROVED** [review]

File: jetson/percep/pcl.cpp (hunk 4 above)
Comment (amverni) [reply jetson/percep/pcl.cpp:465]: You're right, my bad on missing that. Can you just throw in the link to the issue for determining the final value for the x value. I think it is best to just have a "finalize all config file values" issue that we do right before competition. Until then just have local changes since this value will change a bunch throughout the year.

File: jetson/percep/pcl.hpp (hunk 7 above)
Comment (amverni) [reply jetson/percep/pcl.hpp:55]: Link to the issue?

File: jetson/percep/pcl.cpp [hunk 9]
```diff
@@ -455,17 +456,20 @@ shared_ptr<pcl::visualization::PCLVisualizer> PCL::createRGBVisualizer() {
...
+    PassThroughFilter("z", 7000.0);
+    PassThroughFilter("y", 3000.0);
     DownsampleVoxelFilter();
     RANSACSegmentation("remove");
     std::vector<pcl::PointIndices> cluster_indices;
     CPUEuclidianClusterExtraction(cluster_indices);
     std::vector<std::vector<int>> interest_points(cluster_indices.size(), vector<int> (4));
     FindInterestPoints(cluster_indices, interest_points);
-    bearing = FindClearPath(interest_points, viewer);  
+    bearing = FindClearPath(interest_points, viewer);
```
Comment (amverni) [inline jetson/percep/pcl.cpp:472]: In the summary of the "general cleanup" include an item for "remove trailing whitespace" and "ensure all files have a single newline at the end"

**Review by raytitan — APPROVED** [review]

**Review by theonlynick0430 — APPROVED** [review]

File: jetson/percep/pcl.cpp (hunk 9 above)
Comment (mmartin4972) [reply jetson/percep/pcl.cpp:472]: https://github.com/umrover/mrover-workspace/issues/516

Comment (mmartin4972) [reply jetson/percep/pcl.cpp:472]: Y axis filter testing issue: https://github.com/umrover/mrover-workspace/issues/517

File: jetson/percep/pcl.hpp (hunk 7 above)
Comment (mmartin4972) [reply jetson/percep/pcl.hpp:55]: https://github.com/umrover/mrover-workspace/issues/516

File: jetson/percep/pcl.cpp (hunk 4 above)
Comment (amverni) [reply jetson/percep/pcl.cpp:465]: #517

**Review by amverni — APPROVED** [review]
